// etherreq/src/etherreq.js
// src/etherreq.js

export const create = (defaultConfig = {}) => {
  const interceptors = {
    request: [],
    response: [],
  };

  const use = (fulfilled, rejected, type = 'request') => {
    interceptors[type].push({ fulfilled, rejected });
  };

  // ✅ 使用 const 显式声明 dispatchRequest
// etherreq.js
const dispatchRequest = async (config) => {
  // 请求拦截器执行
  for (const interceptor of interceptors.request) {
    config = await interceptor.fulfilled(config);
  }

  let response;

  const { url, method = 'GET', headers = {}, body } = config;
  try {
    const options = {
      method,
      headers,
      body: method !== 'GET' ? JSON.stringify(body) : undefined,
    };

    const res = await fetch(url, options);

    // 先读取文本判断是否是 HTML
    const text = await res.text();

    if (!res.ok) {
      throw new Error(`HTTP 错误: ${res.status} - ${res.statusText}`);
    }

    // 判断是否是 HTML 内容
    if (text.trim().startsWith('<')) {
      throw new Error('服务器返回了 HTML 页面，预期为 JSON 格式');
    }

    const data = JSON.parse(text); // 安全解析

    response = {
      data,
      status: res.status,
      statusText: res.statusText,
      headers: res.headers,
      config,
    };
  } catch (error) {
    for (const interceptor of interceptors.response) {
      if (interceptor.rejected) {
        return interceptor.rejected(error);
      }
    }

    throw error;
  }

  // 响应拦截器执行
  for (const interceptor of interceptors.response) {
    response = await interceptor.fulfilled(response);
  }

  return response;
};

  const instance = (config) => {
    return dispatchRequest({
      ...defaultConfig,
      ...config,
    });
  };

  instance.use = (fulfilled, rejected) => use(fulfilled, rejected, 'request');
  instance.interceptors = {
    request: {
      use: (fulfilled, rejected) => use(fulfilled, rejected, 'request'),
    },
    response: {
      use: (fulfilled, rejected) => use(fulfilled, rejected, 'response'),
    },
  };

  return instance;
};
